src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOp.java

```diff
@@ -1192,7 +1005,8 @@ public Block.Builder lower(Block.Builder b, OpTransformer opT) {
                 } else { // statement body
                     curr.transformBody(bodies().get(i), blocks.get(i).parameters(), opT.andThen((block, op) -> {
                         switch (op) {
-                            case YieldOp yop -> block.op(branch(exit.successor()));
+                            case YieldOp yop when yop.operands().isEmpty() -> block.op(branch(exit.successor()));
+                            case YieldOp yop -> block.op(branch(exit.successor(block.context().getValue(yop.yieldValue()))));
```

Comment:
It might be clearer to test against the concrete switch class e.g., `case YieldOp yop when this instanceof ...`.